scipy/stats/_pagel.py

```diff
@@ -0,0 +1,525 @@
+    # L statistic under H0 would be greater than the observed L statistic
+    p = norm.sf(Lambda)
+    return p
+
+
+def _l_p_exact(L, m, n):
+    '''Calculate the p-value of Page's L exactly'''
+    L, n, k = int(L), int(m), int(n)  # different papers use different symbols
+
+    _pagel_state.set_k(k)
+    try:
+        pmf = _pagel_state.all_pmfs[n][k]
+        return np.sum(pmf[L-_pagel_state.a*n:])
+    except KeyError:
+        return _pagel_state.sf(L, n)
+
+
+class _PageL:
+    '''Class to maintain state of Page's L between `pagel` executions'''
+
+    def __init__(self):
+        '''Fast initialization; doesn't load data from file.'''
+        self.all_pmfs = None
+        self.all_counts = None
+        self.k = None
+        self.a = None
+        self.b = None
+        self.p_l_k_1 = None
+        dir_path = os.path.abspath(os.path.dirname(__file__))
+        self.all_pmfs_file = os.path.join(dir_path, "data_folder",
+                                          "pagel_pmfs.npy")
+        self.all_counts_file = os.path.join(dir_path, "data_folder",
+                                            "pagel_counts.npy")
+
+    def load_pagel_data(self):
+        '''Loads PageL data from file'''
+        # functions for creating these files are at the end of the class
+        self.all_counts = np.load(self.all_counts_file, allow_pickle=True)
+        self.all_pmfs = np.load(self.all_pmfs_file, allow_pickle=True).item()
+
+    def set_k(self, k):
+        '''Generates function to evaluate p(t, k, 1); see [5] Equation 6'''
+
+        if self.k == k:
+            return
+
+        if self.all_counts is None:
+            self.load_pagel_data()
+
+        self.k = k
+        # min L, max L of a row
+        self.a, self.b = (k*(k+1)*(k+2))//6, (k*(k+1)*(2*k+1))//6
+
+        try:
+            counts = self.all_counts[k-1]
+        except IndexError:
+            counts = self.counts_l_k_1(k)
+
+        # See [5] Equation (6)
+        ps = np.array(counts)/np.math.factorial(k)
+
+        def p_l_k_1(l):
+            if l < self.a or l > self.b:  # 0 rows have L < a or L > b
+                return 0
+            return ps[l-self.a]
+
+        self.p_l_k_1 = p_l_k_1
+
+    def pmf(self, l, n):
+        '''Probability mass function of Page's L statistic'''
+        return _pmf_recursive(l, self.k, n, self.a, self.b, self.p_l_k_1)
+
+    def sf(self, l, n):
+        '''Survival function of Page's L statistic'''
+        ps = [self.pmf(l, n) for l in range(l, n*self.b + 1)]
+        return np.sum(ps)
+
+    def whole_pmf(self, k, n):
+        '''Return list of all values of Page's L PMF'''
+        self.set_k(k)
+        ps = [self.pmf(l, n) for l in range(self.a*n, self.b*n + 1)]
+        return ps
+
+    def all_pmf(self):
+        '''Generate all PMFs tabulated in [1]'''
+        data = {}
+        for n in range(2, 13):
+            data[n] = {}
+            for k in range(3, 9):
+                print(n, k)
```

Comment:
Remove this print, or make it more descriptive

Reply:
Yup. That was mostly for debugging. We don't need it now.